com/williamfiset/algorithms/dp/VirtualMachinesScheduler.java

```diff
@@ -0,0 +1,177 @@
+ * <p>5 6 3 Each process has 4 steps and there are 3 VMs. 7 8 5 The first VM runs the 4 processes
+ * with 5, 7, 7, 2 cost respectively and so on. 7 8 3 2 7 6
+ *
+ * <p>0 7 2 7 0 2 2 2 0
+ *
+ * <p>The Algorithm fills an array Costs NxM, where each cell Cost(i, j) contains the minimum total
+ * cost of the complex process until step i when it runs on VM-j. In order for step i to run on
+ * VM-j, the results from i-1 step must already be on VM-j or sent on VM-j including the
+ * communication cost between VMs.
+ *
+ * <p>For the example above the Cost array is the following:
+ *
+ * <p>5 6 3 12 13 8 17 18 11 15 20 17
+ *
+ * <p>This means that the least expensive execution costs 15 and is achieved when the last process
+ * is on the first VM.
+ *
+ * @author Dragatis Nikolas, nikosdraga@gmail.com
+ */
+public class VirtualMachinesScheduler {
+  public static void main(String[] args) throws Exception {
+    double time = System.currentTimeMillis();
+    Scheduler scheduler = new Scheduler();
+
+    scheduler.scanFile(args[0]);
+
+    System.out.println("\nTime: " + (System.currentTimeMillis() - time) / 1000.0 + " seconds.");
+  }
+
+  /**
+   * This is the class that contains the methods for producing the final array of the dynamic
+   * programming after reading and storing the data from a txt file.
+   */
+  public static class Scheduler {
+    private int virtualMachines = 0;
+    private int processes = 0;
+
+    /**
+     * This function is used to scan the txt file, with the data, and initialize the dimensions of
+     * the array with the processes costs and the array with the communication costs.
+     *
+     * @param pathname is the name of the txt file containing the data.
+     * @throws Exception
+     */
+    public void scanFile(String pathname) throws Exception {
+      Scanner scanner = new Scanner(new FileReader(pathname));
+
+      processes = scanner.nextInt();
+      virtualMachines = scanner.nextInt();
+
+      int[][] processesCost = new int[processes][virtualMachines];
+      int[][] communicationsCost = new int[virtualMachines][virtualMachines];
+
+      initializeArrays(scanner, processesCost, processes, virtualMachines);
+
+      initializeArrays(scanner, communicationsCost, virtualMachines, virtualMachines);
+
+      Scheduling(processesCost, communicationsCost);
+    }
+
+    /**
+     * This function is used to produce the final array of the dynamic programming with the results.
+     *
+     * @param processesCost is the array which contains each process cost on each virtual machine.
+     * @param communicationsCost is the array which contains the costs to transfer data from one
+     *     virtual machine to another.
+     */
+    public void Scheduling(int[][] processesCost, int[][] communicationsCost) {
+      int[][] Cost = new int[processes][virtualMachines];
+
+      for (int i = 0; i < virtualMachines; i++) {
+        Cost[0][i] = processesCost[0][i];
+      }
+
+      for (int i = 1; i < processes; i++) {
```

Comment:
Any ideas on the time complexity? O(n^3*log(n))?

Reply:
It should be O(n^2*m), where n=#processes and m=#virtual machines.
The findMin() function is called n*m times and there is another loop inside the function which runs n times.

Reply:
I'm thinking that in Scheduling() function we have 2 loops, one being nested, which run for a total of n*m times. And in there we call the findMin() function which contains another loop which in turn runs m times. So in total it should be n*(m^2) repetitions. But I'm not sure if my thinking is correct or not.